x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -578,17 +597,21 @@ void decodeToken(String token, ActionListener<UserToken> listener) {
     public void invalidateAccessToken(String accessToken, ActionListener<TokensInvalidationResult> listener) {
         ensureEnabled();
         if (Strings.isNullOrEmpty(accessToken)) {
-            listener.onFailure(traceLog("no access token provided", new IllegalArgumentException("access token must be provided")));
+            listener.onFailure(traceLog("invalidate access token", new IllegalArgumentException("access token must be provided")));
         } else {
+            logger.debug("INVALIDATE ACCESS TOKEN");
             maybeStartTokenRemover();
             final Iterator<TimeValue> backoff = DEFAULT_BACKOFF.iterator();
             decodeToken(accessToken, ActionListener.wrap(userToken -> {
                 if (userToken == null) {
-                    listener.onFailure(traceLog("invalidate token", accessToken, malformedTokenException()));
+                    // The chances of a random token string decoding to something that we can read is minimal, so
+                    // we assume that this was a token we have created but is now expired/revoked and deleted
+                    logger.trace("The access token [{}] is expired and already deleted", accessToken);
+                    listener.onResponse(TokensInvalidationResult.emptyResult(RestStatus.NOT_FOUND));
                 } else {
                     indexInvalidation(Collections.singleton(userToken), backoff, "access_token", null, listener);
                 }
-            }, listener::onFailure));
+            }, e -> listener.onFailure(unableToPerformAction(e))));
```

Comment:
Will this change an index related error (unavailable or not exists) to an security exception (401)? Do we need a `isShardNotAvailableException` here?

Reply:
> Will this change an index related error (unavailable or not exists) to an security exception (401)?

No, because `unableToPerformAction` will set the status code to 503. The idea here is that if there is an "expected" reason this failed to decode, we catch it in the `decodeToken` and call the listeners `onResponse()` with `null`. If something unexpected happen, we want to return `503` so that the caller knows that the token might still not be invalidated and should probably try again

Reply:
You are right. The status is service unavailable, not 401.